### PR TITLE
OpenAI provider tutorial

### PR DESCRIPTION
**Pull request by TJaniF**
🖖 

I followed the Weaviate tutorial structure, so the wording is very similar, this might end up being the case for most of the AI provider tutorials 😅

### CODE REVIEW

File: learn/airflow-openai.md [hunk 1]
```diff
@@ -0,0 +1,111 @@
...
+## Step 1: Configure your Astro project
+
+1. Create a new Astro project:
+
+    ```sh
+    $ mkdir astro-openai-tutorial && cd astro-openai-tutorial
+    $ astro dev init
+    ```
+
+2. Add the following install statement to your `requirements.txt` file to install the OpenAI Airflow provider and other supporting packages:
```
Comment (kentdanas) [inline learn/airflow-openai.md:54]:
```suggestion
2. Add the following packages to your `requirements.txt` file to install the OpenAI Airflow provider and other supporting packages:
```

File: learn/airflow-openai.md [hunk 2]
```diff
@@ -0,0 +1,111 @@
...
+
+2. In the Airflow UI, run the `captains_dag` DAG by clicking the play button. Then, provide [Airflow params](airflow-params.md) for:
+
+    - `Question to ask the captain`: The question you want to ask the captains.
+    - `captains_to_ask`: A list of captains you want to ask the question to. Make sure to create one line per captain and to provide at least two names.
+    - `max_tokens_answer`: The maximum number of tokens available for the answer. 
+    - `randomness_of_answer`: The randomness of the answer. The value provided is divided by 10 and given to the `temperature` parameter of the [chat completion endpoint](https://platform.openai.com/docs/guides/text-generation/reproducible-outputs). The scale for the param ranges from 0 to 20, with 0 being the most deterministic and 20 being the most random.
+
+    ![Screenshot of the Airflow UI showing the successful completion of the `captains_dag` DAG in the Grid view with the Graph tab selected. All 8 captains available were selected to be asked the question, which led to 8 mapped task instances of both the `ask_a_captain` and `get_embeddings` task.](/img/tutorials/airflow-openai_dag_completed.png)
```
Comment (kentdanas) [inline learn/airflow-openai.md:101]: This screenshot feels a little out of place to me. Maybe we should put this one up above after the descriptions of the tasks in the Create your DAG section, and then here have a screenshot of the params form?

**Review by kentdanas — APPROVED** [review]
Very minor comment about screenshots. Otherwise this looks great to me!

**Review by TJaniF — COMMENTED** [review]
Thank you @kentdanas ! :) 

@lzdanski and @jwitz ready for you 🖖

File: learn/airflow-openai.md [hunk 3]
```diff
@@ -0,0 +1,113 @@
...
+    openai==0.28.1
+    matplotlib==3.8.1
+    seaborn==0.13.0
+    scikit-learn==1.3.2
+    pandas==1.5.3
+    numpy==1.26.2
+    adjustText==0.8
+    ```
+
+3. To create an [Airflow connection](connections.md) to OpenAI, add the following environment variables to your `.env` file. Make sure to provide you own OpenAI API key.
```
Comment (jwitz) [inline learn/airflow-openai.md:67]: Seeing OpenAI API all the time, while OpenAPI is also a thing, is making my head spin haha

File: learn/airflow-openai.md [hunk 4]
```diff
@@ -0,0 +1,113 @@
...
+
+    ```text
+    AIRFLOW_CONN_OPENAI_DEFAULT='{
+        "conn_type": "openai",
+        "password": "YOUR OPENAI API KEY",
```
Comment (jwitz) [inline learn/airflow-openai.md:72]:
```suggestion
        "password": "<your-openai-api-key>",
```

File: learn/airflow-openai.md (hunk 3 above)
Comment (jwitz) [inline learn/airflow-openai.md:67]:
```suggestion
3. To create an [Airflow connection](connections.md) to OpenAI, add the following environment variables to your `.env` file. Make sure to replace `<your-openai-api-key>` with your own OpenAI API key.
```

File: learn/airflow-openai.md [hunk 5]
```diff
@@ -0,0 +1,113 @@
...
+
+1. In your `dags` folder, create a file called `captains_dag.py`.
+
+2. Copy the following code into the file.
+
+    <CodeBlock language="python">{captains_dag}</CodeBlock>
+
+    This DAG consists of four tasks to make a simple MLOps pipeline.
+
+    ![Screenshot of the Airflow UI showing the successful completion of the `captains_dag` DAG in the Grid view with the Graph tab selected. All 8 captains available were selected to be asked the question, which led to 8 mapped task instances of both the `ask_a_captain` and `get_embeddings` task.](/img/tutorials/airflow-openai_dag_completed.png)
```
Comment (jwitz) [inline learn/airflow-openai.md:86]: I'd move this to after the task descriptions.

File: learn/airflow-openai.md [hunk 6]
```diff
@@ -0,0 +1,113 @@
...
+
+    - The `get_captains_list` task fetches the list of captains that the user wants to ask their question to from the input given via [Airflow params](airflow-params.md).
```
Comment (jwitz) [inline learn/airflow-openai.md:88]:
```suggestion
    - The `get_captains_list` task fetches the list of Star Trek captains you want to ask your question to. You'll provide the list of captains when you run the DAG through [Airflow params](airflow-params.md).
```
I thought "Captains" was an AI term until I got to the end of the page! (Yes I did skip reviewing the DAG 🤧)

File: learn/airflow-openai.md [hunk 7]
```diff
@@ -0,0 +1,113 @@
...
+    - The `ask_a_captain` task uses the [OpenAIHook](https://airflow.apache.org/docs/apache-airflow-providers-openai/stable/_api/airflow/providers/openai/hooks/openai/index.html) to connect to the OpenAI API and use the [chat completion endpoint](https://platform.openai.com/docs/guides/text-generation/chat-completions-api) to leverage the `gpt-3-turbo` model to generate answers to the question provided by the user. This task is [dynamically mapped](dynamic-tasks.md) over the list of captains to generate one dynamically mapped task instance per captain.
```
Comment (jwitz) [inline learn/airflow-openai.md:89]:
```suggestion
    - The `ask_a_captain` task uses the [OpenAIHook](https://airflow.apache.org/docs/apache-airflow-providers-openai/stable/_api/airflow/providers/openai/hooks/openai/index.html) to connect to the OpenAI API. It then uses the [chat completion endpoint](https://platform.openai.com/docs/guides/text-generation/chat-completions-api) generate answers to the question you provide. This task is [dynamically mapped](dynamic-tasks.md) over the list of captains to generate one dynamically mapped task instance per captain.
```

File: learn/airflow-openai.md [hunk 8]
```diff
@@ -0,0 +1,113 @@
...
+    - The `get_embeddings` task is defined using the [OpenAIEmbeddingOperator](https://airflow.apache.org/docs/apache-airflow-providers-openai/stable/operators/openai.html) to generate vector embeddings of the answers generated by the upstream `ask_a_captain` task. This task is dynamically mapped over the list of answers to retrieve one set of embeddings per answer. This pattern allows for efficient parallelization of the vector embedding generation.
+    - The `plot_embeddings` task takes the embeddings created by the upstream task and performs dimensionality reduction using [PCA](https://scikit-learn.org/stable/modules/generated/sklearn.decomposition.PCA.html) to plot the embeddings in two dimensions. 
+
+## Step 3: Run your DAG
+
+1. Run `astro dev start` in your Astro project to start Airflow and open the Airflow UI at `localhost:8080`.
```
Comment (jwitz) [inline learn/airflow-openai.md:95]:
```suggestion
1. Run `astro dev start` in your Astro project to start Airflow, then open the Airflow UI at `localhost:8080`.
```

File: learn/airflow-openai.md [hunk 9]
```diff
@@ -0,0 +1,113 @@
...
+
+2. In the Airflow UI, run the `captains_dag` DAG by clicking the play button. Then, provide [Airflow params](airflow-params.md) for:
+
+    - `Question to ask the captain`: The question you want to ask the captains.
+    - `captains_to_ask`: A list of captains you want to ask the question to. Make sure to create one line per captain and to provide at least two names.
```
Comment (jwitz) [inline learn/airflow-openai.md:100]:
```suggestion
    - `captains_to_ask`: A list of Star Trek captains you want to ask the question to. Make sure to create one line per captain and to provide at least two names.
```

File: code-samples/dags/airflow-openai/captains_dag.py [hunk 10]
```diff
@@ -0,0 +1,215 @@
+"""
+## Ask questions to Star Trek captains using OpenAI's LLMs, embed and visualize the results
+
+This DAG shows how to use the OpenAI Airflow provider to interact with the OpenAI API.
+The DAG asks a question to a list of Star Trek captains given using Airflow params,
```
Comment (jwitz) [inline code-samples/dags/airflow-openai/captains_dag.py:5]:
```suggestion
The DAG asks a question to a list of Star Trek captains based on values you provide via Airflow params,
```

**Review by jwitz — APPROVED** [review]
LGTM! Another banger use case 🔥

File: learn/airflow-openai.md (hunk 3 above)
Comment (TJaniF) [reply learn/airflow-openai.md]: Hahaha, I had to say "OpenAI API" a couple of times in a talk and a webinar and it was tough not to stumble. 😂

File: learn/airflow-openai.md (hunk 6 above)
Comment (TJaniF) [reply learn/airflow-openai.md]: 🤣  I'll see if I can add a 🖖 emoji somewhere

File: learn/airflow-openai.md [hunk 11]
```diff
@@ -0,0 +1,113 @@
...
+
+## Why use Airflow with OpenAI?
+
+OpenAI offers a variety of powerful model endpoints for different tasks like text or image generation, vector embeddings or translation tasks. These models are used in countless applications, both user-facing, such as chatbots, and internal, for example providing insight into internal knowledge base content.
+
+Integrating OpenAI with Airflow into one end-to-end machine learning pipeline allows you to:
+
+- Use Airflow's [data-driven scheduling](airflow-datasets.md) to run operations using OpenAI model endpoints based on upstream events in your data ecosystem, such as when new user input is ingested or a new dataset is available.
+- Send several requests to a model endpoint in parallel based on upstream events in your data ecosystem or user input via [Airflow params](airflow-params.md).
+- Add Airflow features like [retries](rerunning-dags.md#automatically-retry-tasks) and [alerts](error-notifications-in-airflow.md) to your OpenAI operations. For example for handling or monitoring service outages.
```
[the remark this answers is not in the document]
Comment (TJaniF) [reply learn/airflow-openai.md]: I changed this sentence a bit to be more specific, let me know if it works for you :)

**Review by TJaniF — COMMENTED** [review]
Thank you @jwitz ! 

I had one sentence that I redid based on your feedback, other than that accepted everything :)

**Pull request merged**